public/components/main/main.tsx

```diff
@@ -64,11 +79,182 @@ export class Main extends React.Component<RouterHomeProps, any> {
     pageSizeOptions: [8, 10, 13],
   };
 
+  GenerateReportLoadingModal() {
+    const [isModalVisible, setIsModalVisible] = useState(true);
+
+    const closeModal = () => setIsModalVisible(false);
+    const showModal = () => setIsModalVisible(true);
+
+    return (
+      <div>
+        <EuiOverlayMask>
+          <EuiModal onClose={closeModal}>
+            <EuiModalHeader>
+              <EuiTitle>
+                <EuiText textAlign="right">
+                  <h2>Generating report</h2>
+                </EuiText>
+              </EuiTitle>
+            </EuiModalHeader>
+            <EuiModalBody>
+              <EuiText>Preparing your file for download...</EuiText>
+              <EuiSpacer />
+              <EuiFlexGroup justifyContent="spaceAround" alignItems="center">
+                <EuiFlexItem>
+                  <EuiLoadingSpinner size="xl" />
+                </EuiFlexItem>
+              </EuiFlexGroup>
+            </EuiModalBody>
+          </EuiModal>
+        </EuiOverlayMask>
+      </div>
+    );
+  }
+
+  readStreamToFile = async (stream: string) => {
+    var link = document.createElement('a');
+    var fileFormatPrefix = getFileFormatPrefix(
+      this.state.generateReportFileFormat
+    );
+    var url = fileFormatPrefix + stream;
+    if (typeof link.download == 'string') {
+      if (fileFormatPrefix.includes('pdf')) {
+        link.download = this.state.generateReportFilename.substring(
+          0,
+          this.state.generateReportFilename.length - 4
+        );
+      } else {
+        link.download = this.state.generateReportFilename;
+      }
+      link.href = url;
+      document.body.appendChild(link);
+      link.click();
+      await this.updateReportsTableContent();
+      document.body.removeChild(link);
+    } else {
+      window.open(url, '_blank');
+    }
```

Comment:
could you try return early to avoid nest-if 

```
if (typeof link.download !== string) {
  window.open(url, '_blank');
  return;
}

...


```

Reply:
Will change